### PR TITLE
Replace SYCL backend `reduce_by_segment` implementation with reduce-then-scan call

### PR DESCRIPTION
**Pull request by mmichel11**
## Summary
This PR implements a SYCL backend `reduce_by_segment` by using higher level calls to reduce-then-scan along with new specialty functors to achieve a segmented reduction. This PR is an initial step of porting the implementation to reduce-then-scan with optimization likely to follow. Future efforts may include additional modification to reduce-then-scan kernels.

Performance improves for all input sizes. For small inputs, we see 3-5x improvements and for very large sizes ~1.25x on GPU Series Max 1550. Please contact me if you would like to see performance data.

## Description of changes
- The SYCL `reduce_by_segment` implementation that was previously handwritten is replaced by a higher level call to our reduce-then-scan kernels. Several new callback functors for the reduce-then-scan kernel have been made to achieve this operation.
- `reduce_by_segment.pass` was encountering linker crashes due to the large number of test cases being compiled growing past the maximum size of the binary's data region. SYCL testing has been trimmed down with regards to USM device and shared testing which resolves this issue. Instead of running each test with a device and shared USM allocation, every other test switches the USM type.
- `ONEDPL_WORKAROUND_FOR_IGPU_64BIT_REDUCTION` has been removed as the SYCL implementation has been replaced, and we are no longer impacted by this issue.
- The legacy `reduce_by_segment` implementation is used as a fallback for when the sub-group size, device, and trivial copyability constraints cannot be satisfied.

## Future work
Future efforts on `reduce_by_segment` may built on top of this implementation and the reduce-then-scan kernels to better handle first and last element cases.

### CODE REVIEW

**Comment by mmichel11** [comment]
I have made some design changes based on offline discussion. There is quite a bit of code movement that has happened, so here is a summary of the recently made changes:

- An iterator-based `__pattern_reduce_by_segment` has been added to `algorithm_impl_hetero.h`. Previously, we just had a range-based version. This resolves the issue of calling range-based patterns from iterator-based algorithms.
- The fallback `reduce_by_segment` implementation based on high-level `copy_if` and `parallel_for` calls has been moved down a level from `algorithm_ranges_impl_hetero.h` to `dpcpp/parallel_backend_sycl.h` so that we can fallback on this implementation when reduce-then-scan cannot be used in `__parallel_reduce_by_segment`. This pattern has been implemented synchronously as parallel pattern calls cannot currently depend on each other.
- Due to observed performance issues for compilers prior to icpx 2025.0, the reduce-then-scan path must be disabled in this case. The known-identity based implementation has been added back to avoid introducing a performance regression for `icpx 2024.2.1` and prior. It has been moved into `dpcpp/parallel_backend_sycl_reduce_by_segment.h` so the implementations are not split across several directories.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:2345]: This is perhaps just a preference thing for code organization, but perhaps we can name this implementation and put it in its own function, that way this function can just be the algorithm selection, and it may be more clear.

Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:2345]: I think this is a good idea and will make this change.

Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I have made this change. The two fallback paths I have named to `__parallel_reduce_by_segment_fallback` with a true_type / false_type parameter for whether there is a known identity.

I do not like that one fallback path is in `parallel_backend_sycl.h` and the other is in `parallel_backend_sycl_reduce_by_segment.h`. However, I did this since I did not want to add both fallback paths into `parallel_backend_sycl_reduce_by_segment.h` as it would require forward declarations for the other parallel algorithms calls made, and I also did not want to add another 400+ lines to `parallel_backend_sycl.h`

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h:68]: I confirmed in an outside editor that the only changes to this from the previous location are only cosmetic.  I'm not looking deeply into this files changes otherwise since it has already been reviewed and it was just moved.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:814]: In other reduce_then_scan based API implementations, we are able to reuse the `..._reduce_input` generator inside the `..._scan_input`, but I can see why we may not want to do that here to consolidate branches.  It would be nice to rely on the same code to prevent redundancy and better show connection here.  I was thinking about ways to do it (perhaps with helper functions called by the call operators of each) but it may not be worth it.

I'll let you decide on this.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:848]: add a comment here that this is to handle the `__id==0` case.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:854]: It may be possible to avoid the `__id == 0` case, in a similar way to unique.  It is a little more complicated because we would need to set up the carry-in appropriately, but I think its possible and could provide some branch avoiding (and tuple shrinking) gains in the helpers.  
If you think its possible to do this, lets leave it as an issue to be explored in a follow up.

**Review by adamfidel — COMMENTED** [review]
First pass of comments. I have looked at primarily the fallback algorithms and intend to focus on the reduce-then-scan implementation next.

File: test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp
Comment (mmichel11) [reply test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp:361]: Yes, every test is alternated with testing USM shared vs. USM device memory instead of testing both. Prior to doing this, the data section of the executable grew far too large resulting in a linker error, " relocation truncated to fit...".

From looking the error message up, this could be avoided by using special compilation flags to use 64-bit relative addressing, but I think trimming the test down a bit makes the most sense.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:915]: Good point. I added a comment explaining that this is done to output the first key in each segment along with the special handling of the first and last element.

File: test/general/implementation_details/device_copyable.pass.cpp
Comment (mmichel11) [reply test/general/implementation_details/device_copyable.pass.cpp]: I switched to use `binary_op_device_copyable`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:817]: I took a look at doing this. It is possible to add some helper utilities shared between the reduce / scan input generation, but I do not think it improves the simplifies the code too much unless it is written in a way that has some more conditional checks which I want to avoid.

With the small differences in each branch in `__gen_red_by_seg_scan_input`, I think it is clearest to leave it as is.

**Review by danhoeflinger — COMMENTED** [review]
At a high level I agree with the changes in this PR, but there are still a few remaining nit picks outstanding.  

I have run out of time before my time off to get into the small details like sizes of types and forwarding of references, things like that.  The clang format suggestions can be ignored as of now.

So, I wont hit approve officially but I think this is very close and trust @adamfidel / others to be able to get it across the finish line and have no objections to merging with another approval.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1]: Thanks for the feedback. I have added some more explanations and a small example showing how these functors achieve a the reduce_by_segment operation.

File: test/general/implementation_details/device_copyable.pass.cpp
Comment (mmichel11) [reply test/general/implementation_details/device_copyable.pass.cpp]: I have made this change.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1]: Looks good.  Thanks!

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h]: Good point, it is not needed and I have removed it.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h]: I changed this occurrence and a few lines below in the inclusive scan call as well which was a similar case.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_by_segment.h]: I changed this type to be `std::uin32_t`

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:42]: I took another look at adding all of the implementations to the reduce-by-segment header. It works, but there are a few issues that I think make it cleaner to leave it as is:

- `__parallel_copy_if` has default template / function arguments and providing a forward declaration in `parallel_backend_reduce_by_segment.h` will require having these defaults here and not in `parallel_backend_sycl.h`. This is confusing for code readability in my opinion as the defaults will be in this new header that is unrelated to copy_if itself.
- `__get_zeroth_element` requires a full definition of the struct, so we would need to define a similar struct or move this callback somewhere else. Currently, all of the reduce_then_scan callback functors are in `parallel_backend_sycl.h`
- We have an additional copyright on the fallback implementation currently in the new header. Having only this implementation hear clearly indicates where the copyright applies.

If there are no objections, I think the current approach is best.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:857]: I agree and have opened https://github.com/oneapi-src/oneDPL/issues/1958.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:840]: Since the expression `_ValueType{__in_vals[__id]}` appears in every branch, would it make sense to hoist the copy before the branches and `std::move` the value into each tuple?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:2343]:
```suggestion
//------------------------------------------------------------------------
// reduce_by_segment - sync pattern
//
// TODO: The non-identity fallback path of reduce-by-segment must currently be implemented synchronously due to the
// inability to create event dependency chains across separate parallel pattern calls. If we ever add support for
// cross parallel pattern dependencies, then we can implement this as an async pattern.
//------------------------------------------------------------------------
```

Suggest that we move the comment about the synchronousity of the fallback up because its current placements looks like it's commenting on `__reduce1_wrapper` instead of the pattern as a whole.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:957]: This looks to be assigning a reference from a tuple. I know we have hit snags in the past with references and tuples. Is this safe to do in this instance or do we need to cast to the value type (i.e., `_ValType{__current_value}`)?

**Review by adamfidel — COMMENTED** [review]
A few more comments. 

I think overall the approach is good and I don't mind approving once these few small comments have been addressed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I made this change by adding a new variable. Instead of copying and moving it into the tuple I did what is being done with the keys and pass it as a const reference instead.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:958]: I think `__current_value` is already the correct type without references. This is because its type, `_ValType` is `using _ValType = oneapi::dpl::__internal::__value_t<decltype(__out_values)>;` and the references should already be removed.

Does this make sense?

**Review by adamfidel — APPROVED** [review]
LGTM overall. The `clang-format` changes can be ignored in my opinion.

**Comment by mmichel11** [comment]
I have fixed a compilation issue identified in CI. With some configurations, specifying the full namespace `oneapi::dpl::__ranges` when calling `__get_sycl_range` is needed to avoid compilation errors.

I am rerunning CI to make sure there are no more issues.

**Review by adamfidel — APPROVED** [review]
LGTM after the recent changes to fix issues discovered in CI.

**Review by danhoeflinger — APPROVED** [review]
LGTM after minor changes.  Thanks and sorry for the delay

**Pull request merged**
